_drafts/debug-monitor.md

```diff
@@ -0,0 +1,676 @@
+  const bool is_halt_dbg_evt = (*dfsr & dfsr_halt_evt_bitmask);
+
+  EXAMPLE_LOG("DebugMonitor Exception");
+
+  EXAMPLE_LOG("DEMCR: 0x%08x", *demcr);
+  EXAMPLE_LOG("DFSR:  0x%08x (bkpt=%d, halt=%d, dwt=%d)", *dfsr,
+              (int)is_bkpt_dbg_evt, (int)is_halt_dbg_evt,
+              (int)is_dwt_dbg_evt);
+
+  EXAMPLE_LOG("Register Dump");
+  EXAMPLE_LOG(" r0  =0x%08x", frame->r0);
+  EXAMPLE_LOG(" r1  =0x%08x", frame->r1);
+  EXAMPLE_LOG(" r2  =0x%08x", frame->r2);
+  EXAMPLE_LOG(" r3  =0x%08x", frame->r3);
+  EXAMPLE_LOG(" r12 =0x%08x", frame->r12);
+  EXAMPLE_LOG(" lr  =0x%08x", frame->lr);
+  EXAMPLE_LOG(" pc  =0x%08x", frame->return_address);
+  EXAMPLE_LOG(" xpsr=0x%08x", frame->xpsr);
+
+  // ...
+}
+```
+
+#### Triggering a breakpoint
+
+To test things out we can create a super simple CLI command that issues a breakpoint instruction:
+
+```c
+static int prv_issue_breakpoint(int argc, char *argv[]) {
+  __asm("bkpt 1");
+  return 0;
+}
+```
+
+In one terminal you will need to start a gdbserver:
+
+```bash
+$ JLinkGDBServer  -if swd -device nRF52840_xxAA  -nogui
+```
+
+Then you need to compile the example application and flash it
+
+```bash
+$ cd ${INTERRUPT_REPO}/example/debugmon/
+$ make
+Compiling src/debug_monitor_exception.c
+[...]
+Linking library
+Generated build/nrf52.elf
+$ arm-none-eabi-gdb-py --eval-command="target remote localhost:2331"  --ex="mon reset" --ex="load"
+--ex="mon reset"  --se=build/nrf52.elf
+(gdb)
+```
+
+From GDB. we will disable halting debug so the program will continue running:
+
+```bash
+(gdb) set *(uint32_t*)0xE000EDF0=(0xA05F<<16)
+```
+
+In a final terminal we will open a serial connection using miniterm:
+
+```bash
+$ miniterm.py - 115200 --raw
+--- Available ports:
+...
+---  4: /dev/cu.usbmodem0006838593731 'J-Link'
+--- Enter port index or full name: 4
+--- Miniterm on /dev/cu.usbmodem0006838593731  115200,8,N,1 ---
+--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
+shell> debug_mon_en
+Monitor Mode Debug Enabled!
+shell> bkpt
+DebugMonitor Exception
+DEMCR: 0x01010000
+DFSR:  0x00000002 (bkpt=1, halt=0, dwt=0)
+Register Dump
+ r0  =0x00000001
+ r1  =0x20003310
+ r2  =0x73734900
+ r3  =0x000009d5
+ r12 =0xffffffff
+ lr  =0x0000094b
+ pc  =0x000009d4
+ xpsr=0x210e0000
+Breakpoint Detected, Awaiting 'c' or 's'
+```
+
+As expected we see bit 1 in `DFSR` is set indicating a breakpoint debug event was generated.
+We can use GDB to inspect the instruction located at`0x000009d4` and confirm it is indeed a breakpoint
+instruction:
+
+```
+(gdb) x/i 0x000009d4
+  0x9d4 <prv_issue_breakpoint>:	bkpt	0x0001
+```
+
+We can also dump the instruction and confirm it matches the BKPT instruction encoding (`0xbe..`):
+
+![]({% img_url breakpoint/bkpt-instruction-cortex-m.png %})
+
+```
+(gdb) p/x *(uint16_t*)0x000009d4
+$4 = 0xbe01
+```
+
+### Programmatically installing Breakpoints with the FPB
+
+For the NRF52, we can install a breakpoint using the following logic:
```

Comment:
It would be nice to add a paragraph here on how the FP_COMP register works. No need to go into huge details, but 1-2 sentences would be good.